### DISCUSSION TITLE
[Modular State Management] Best way to "view" into an element of an array?

### DISCUSSION THREAD

**Discussion by eivindml** [pointfreeco/episode-code-samples#87] [Q&A]
Learning a lot from the modular state management series. I have a question with regards to "viewing" into State and Action discussed in episode [#74](https://github.com/pointfreeco/episode-code-samples/tree/master/0074-modular-state-management-view-actions). 

What I'm trying to do is this:
I have an list of projects (`ProjectsView`), that I can click on and get more details about a single project.
And I want to achieve so that the `ProjectView` only have access to that one project (`store: Store<Project, ProjectAction>`).

```swift
struct AppState {
  var projects: [Project] = []
}
```
I can do this to pass down the value.
```swift
struct ProjectsView: View {
  @ObservedObject var store: Store<[Project], ProjectsViewAction>
  
  public var body: some View {
    NavigationView {
      List {
        ForEach(store.value.indices) { index in
          NavigationLink(
            destination: ProjectView(
              store: self.store.view(
                value: { $0[index] },
                action: { .project($0) }
              )
            )
          ) { Text(self.store.value[index].name) }
        }
      }
    }
  }
}
```
But then we don't have enough information to pullback from a single project to the projects array. `pullback(projectReducer, value: \.project, action: \.project)`.

My solution so far is to add a project property in `AppState` and update the corresponding element in the array when this element is updated:
```swift
struct AppState {
  var project: Project? = nil {
    didSet {
      guard
        let index = projects.firstIndex(where: { $0.id == oldValue?.id }),
        let project = oldValue else { return }
        projects[index] = project
    }
  }
  var projects: [Project] = []
}
```

But this doesn't feel quite right. **Is there any better solutions to achieve this?** It seems like a common case to go from a list of things and view into one single element of that list. So would appreciate some input. 🙏🏼

**Comment by eivindml** [comment]
Then the reducer also has to work on an optional, which I wish it didn't.

```swift
public func projectReducer(state: inout Project?, action: ProjectAction) -> Effect {
  switch action {
    
  case .increaseBy(let value):
    state?.time += value
    return {}
  }
}
```

**Comment by AlexisQapa** [comment]
I got the same kind of issue while working with dictionaries or incompatible states represented by an enum.

**Comment by eivindml** [comment]
Hey 👋🏻 @mbrandonw or @stephencelis. Maybe one of you have some feedback with regards to this? Would really appreciate it. 🙏🏼

**Comment by mbrandonw** [comment]
Hey there, I can definitely nudge you in the right direction. 

If I understand correctly, you need the idea of having a list of views that each have their own local reducer, but you want to pull that back to a reducer that works on the entire list of state.

You want the concept of an "indexed reducer", i.e. a higher-order reducer that can lift an existing reducer up to one that operates on an array of states. There are a few ways you can do accomplish, but one way we've had success with has the following form:

```swift
public struct Indexed<Value> {
    public var index: Int
    public var value: Value

    public init(index: Int, value: Value) {
        self.index = index
        self.value = value
    }
}

func indexed<State, Action, GlobalState, GlobalAction>(
  reducer: Reducer<State, Action>,
  _ stateKeyPath: WritableKeyPath<GlobalState, [State]>,
  _ actionKeyPath: WritableKeyPath<GlobalAction, Indexed<Action>?>
) -> Reducer<GlobalState, GlobalAction> {
  fatalError("Exercise left to the reader 😀")
}
```

Then you could use it like this:

```swift
struct AppState {
  var projects: [Project]
}
enum AppAction {
  var projects(Indexed<ProjectAction>)
}

let appReducer = indexed(reducer: projectReducer, \.projects, \.projects)
```

That's the basic idea, there are still a few small blanks to fill in.

We want to talk about these more powerful higher-order reducers soon, but hopefully this can get you started!

**Comment by liscio** [comment]
I am working on something that took me down this path as well, and hit a nasty issue while integrating this into a project that I’m working on.

Building upon the example given above, consider the following:

``` swift
struct ProjectsOverviewView: View {
    @ObservedObject var store: Store<ProjectsOverviewState, ProjectsOverviewAction>
    
	public var body: some View {
        NavigationView {
            List {
                ForEach(store.value.projects.indices) { index in
                    HStack {
                        ProjectCellView(store: self.store.view(
                          value: { $0.projects[index] },
                          action: { .projectCell(Indexed(index: index, value: $0)) })
                        )

                        Spacer()

                        Button(“Delete”) {
                          self.store.send(.deleteProjectAt(index))
                        }
                    }
                }
            }
        }
    }
}
```


Specifically, there is a list of `ProjectCellView` instances that represent rows in a list, with an inline button that triggers deletion. The `ProjectsOverviewState` maintains the array of `projects`

Deletion will cause one of the `Project`s to be removed from the array, similar to how this is done in the `FavoritePrimesView` from the series.

Upon executing `send(.deleteProjects($0))`, the observation that is created in the `Store`’s `view` function is triggered immediately—once for each of the `Store`s belonging to the `ProjectCellView`s. 

The trouble is that there is now one less item. The last `Store` created by `Store.view` will still receive the update and invoke the closure passed into the `value` parameter. Boom—that index no longer exists!

The solution I came up with is rather gross, unfortunately:

```swift
public final class Store<Value, Action>: ObservableObject {
	// Existing implementation

    // Cancel any observations on the parent store
	public func cancelUpdates() { self.cancellable = nil }
}
```

It requires that we expose an aspect of the implementation that _absolutely_ should not leak outside the `Store` class.

Furthermore, it requires that I somehow maintain references to the individual `Store`s that are created for child elements so that `cancelUpdates()` is invoked before the `.deleteProjects` action is sent. Something like this:

```swift
struct ProjectsOverviewView: View {
    @ObservedObject var store: Store<ProjectsOverviewState, ProjectsOverviewAction>
    
    public var body: some View {
        NavigationView {
            List {
                ForEach(store.value.projects.indices) { index in
                    makeCellForProject(at: index)
                }
            }
        }
    }
    
    private func makeCellForProject(at index: Int) -> some View {
        let localStore = self.store.view(
            value: { $0.projects[index] },
            action: { .projectCell(Indexed(index: index, value: $0)) }
        )
        
        return HStack {
            ProjectCellView(store: localStore)

            Spacer()

            Button(“Delete Project”) {
                localStore.cancelUpdate()
                self.store.send(.deleteProjectAt(index))
            }
        }

    }
}
```

So this fixes _my_ specific issue, however it only works because my design calls for an inline delete button such that the child created by `Store.view` can be made available within the same scope.

For example, consider if I used the `onDelete` handler, instead:

```swift
struct ProjectsOverviewView: View {
    @ObservedObject var store: Store<ProjectsOverviewState, ProjectsOverviewAction>
    
    public var body: some View {
        NavigationView {
            List {
                ForEach(store.value.projects.indices) { index in
                    ProjectCellView(
                        store: self.store.view(
                            value: { $0.projects[index] },
                            action: { .projectCell(Indexed(index: index, value: $0)) }
                        )
                    )
                }
            }
        }
        .onDelete { self.store.send(.deleteProjectsAt($0)) }
    }
}
```

In _this_ case, there is no easy way to maintain references to the `Store` objects created by the `Store.view` method, and connect them to the `delete` operation—or _any_ mutating operation that would affect `projects` in any way that should invalidate individual entries.

So I’m thinking that—in addition to an `indexed` reducer, we probably also need some kind of `indexedView` that would provide us with a mechanism to do one of the following:

1. Notify the `Store`s created by `indexedView` that an observation should be torn down (or temporarily ignored).
2. Have the `Store`s created by `indexedView` check whether the index/item still exists before notifying a child `Store` about the change.
3. A completely different (and far better?) solution that I have failed to consider.

On top of the above concerns, there is also the following possibility—what if I structured things such that the child view housed the Delete button, and I wanted to “bubble up” a `deleteProject` action in the `ProjectCellAction` enum, relying on `Indexed<ProjectCellAction>` to tell me what index is to be deleted in the reducer for `ProjectsOverviewView`?

Lots to consider here! And _so many_ follow-on questions I’m holding back!!

@mbrandonw—is this the sort of thing that you’re planning to cover in a future episode?

(Apologies for the lengthy post, but I’m especially captivated by this series as it speaks directly to the issues I’ve had with structuring some intensely complicated UI that I’m working on right now. Keep up the excellent work!)

**Comment by eivindml** [comment]
@liscio If I understand you correctly, this is a general issue with iterating over an array in SwiftUI, using indicies. You can iterate over each project instead.

```swift
ForEach(store.value) { project in
          NavigationLink(
            destination:
            ProjectView(
              store: self.store.view(
                value: { _ in project },
                action: { .projects(Indexed(index: self.store.value.firstIndex(where: { $0.id == project.id})!, value: $0)) }
              )
            )
          ) {
            VStack(alignment: .leading) {
              Text(project.name)
              Text("\(project.time)")
            }
          }
        }
        .onDelete { self.store.send(.projects(.remove($0))) }
```


But I'm struggeling with the `func indexed()` implementation. Do you have an implementation for that you could share with me? Would appreciate it. :) 

**Edit:**

This is my current implementation. It works, but I feel like I only know 60% what I'm doing here. I just modified the pullback method. I think it's KeyPath's that I don't have a firm grasp of, that makes this a bit difficult to understand. Think I will go back and re-watch some episodes.

```swift
public func indexed<State, Action, GlobalState, GlobalAction>(
  reducer: @escaping Reducer<State, Action>,
  _ stateKeyPath: WritableKeyPath<GlobalState, [State]>,
  _ actionKeyPath: WritableKeyPath<GlobalAction, Indexed<Action>?>
) -> Reducer<GlobalState, GlobalAction> {
  return { globalValue, globalAction in
    guard let localAction = globalAction[keyPath: actionKeyPath] else { return [] }
    let index = localAction.index
    let localEffects = reducer(&globalValue[keyPath: stateKeyPath][index], localAction.value)
    return localEffects.map { localEffect in
      { callback in
        localEffect { localAction in
          var globalAction = globalAction
          globalAction[keyPath: actionKeyPath] = Indexed(index: index, value: localAction)
          callback(globalAction)
        }
      }
    }
  }
}
```

**Comment by mbrandonw** [comment]
> @mbrandonw—is this the sort of thing that you’re planning to cover in a future episode?
> 
> (Apologies for the lengthy post, but I’m especially captivated by this series as it speaks directly to the issues I’ve had with structuring some intensely complicated UI that I’m working on right now. Keep up the excellent work!)

@liscio It's awesome to see that you've dug so deep into these topics, and yes we are absolutely going to cover them at some point. I will say that working with indices in SwiftUI is definitely thorny, where old views appear to get updates even after they should have been destroyed. @stephencelis  submitted radars, but who knows if it's expected SwiftUI behavior or not 😔

But the solution that @eivindml posted is a good approach, and is closely related to the ideas we will be exploring at some point.

> But I'm struggeling with the func indexed() implementation. Do you have an implementation for that you could share with me? Would appreciate it. :)

@eivindml This implementation is exactly right 😀 It runs the reducer on the state at a particular index, and then the local effects are transformed into global effects by running each local effect, getting the local action, and putting it back into the global action at a specific index.

It's a mouthful, but it's right!

**Comment by mbrandonw** [comment]
> Then the reducer also has to work on an optional, which I wish it didn't.
> 
> ```swift
> public func projectReducer(state: inout Project?, action: ProjectAction) -> Effect {
>   switch action {
>     
>   case .increaseBy(let value):
>     state?.time += value
>     return {}
>   }
> }
> ```

@eivindml I didn't see this the first time I read this thread, but there's a nice solution to that too. You can define another higher-order reducer for lifting a reducer to optionals:

```swift
func optionalize<State, Action>(
  reducer: @escaping (inout State, Action) -> Effect<Action>
) -> (inout State?, Action) -> Effect<Action> {
  fatalError("Unimplemented")
}
```

This allows your reducer to work on honest, non-optional state while still allowing it to be pulled back to global state that has optionals.

Might come in handy!

(p.s. the ergonomics of higher-order reducers can get pretty gnarly with free functions, so soon we plan to do an ergonomics episodes upgrading reducers to a proper `struct Reducer { reducer: (inout State, Action) -> [Effect<Action>] }` type. If you are doing a bunch of higher-order reducer work you may want to try it out)

**Comment by liscio** [comment]
> this is a general issue with iterating over an array in SwiftUI, using indicies. You can iterate over each project instead.

@elvindml The point I failed to make with my ill-conceived example was that the "child view" often uses a derived state (e.g. if we generate a `ProjectCellState`) for each cell. In that case, we would have to propagate the identity (i.e. the `id` in the `Identified` protocol) down into the cell state to keep them stable whenever SwiftUI executes your `ForEach` block.

But you also have to be careful about how you push changes back into individual projects—the straightforward approach that you use in the `indexed` reducer (which matches my own—great minds think alike!) would not be ideal in this updated scenario. It would require that the top-level Store re-generates an array of `[ProjectCellState]` every time an action is executed—just to modify a single item!

From your implementation:

```swift
public func indexed<State, Action, GlobalState, GlobalAction>(
  reducer: @escaping Reducer<State, Action>,
  _ stateKeyPath: WritableKeyPath<GlobalState, [State]>,
  _ actionKeyPath: WritableKeyPath<GlobalAction, Indexed<Action>?>
) -> Reducer<GlobalState, GlobalAction> {
  return { globalValue, globalAction in
    guard let localAction = globalAction[keyPath: actionKeyPath] else { return [] }
    let index = localAction.index
    // NOTE: globalValue[keyPath: stateKeyPath] would need to generate a new `[ProjectCellState]` here for every action.
    let localEffects = reducer(&globalValue[keyPath: stateKeyPath][index], localAction.value)
    return localEffects.map { localEffect in
      { callback in
        localEffect { localAction in
          var globalAction = globalAction
          globalAction[keyPath: actionKeyPath] = Indexed(index: index, value: localAction)
          callback(globalAction)
        }
      }
    }
  }
}
```

To fix this, the `indexed` reducer could be extended with a getter/setter in place of the `stateKeyPath` to avoid that problem.

You could _also_ create an `identified` reducer that works similarly, so that views are not required to repeat the logic for reading/writing values based on their `id`.

Hopefully the above makes sense and/or helps in some way. 😄

**Comment by liscio** [comment]
>  It's awesome to see that you've dug so deep into these topics, and yes we are absolutely going to cover them at some point. 

Like I said above—this hits me at just the right time as I'm working on new features that must be isolated from the rest of a larger, legacy app. I can't really afford to import the entirety of the domain/business logic, and having "global" (to the sub-component) state gives me a clearer boundary that lets me interface with what I've already built.

> I will say that working with indices in SwiftUI is definitely thorny, where old views appear to get updates even after they should have been destroyed. @stephencelis submitted radars, but who knows if it's expected SwiftUI behavior or not 😔

@mbrandonw I'm not sure on the specifics of @stephencelis' radar(s), but I would fully expect SwiftUI to continue calling the `body` of subviews as it renders certain animations.

Also, views don't really have a concept of "lifetime". While a view might appear to be "omitted" by an `if`/`else` block within a builder function, it still exists inside the struct that wraps the conditional. So if the structs are copied around, I would not be surprised if their `@ObservedObject`s are still kept alive after the view disappears from the screen.

In the case that I raised: the deletion happens, causing the parent `Store` to get updated, which executes the `sink` immediately, and then the `LocalValue` is no longer found in the array. At the time of deletion, the view is definitely still on the screen and this is totally expected behavior (now that I understand it more.)

My fix for the problem was to create an `indexedView` on the `Store` that has a `toLocalValue` closure that also takes an index, but returns an optional `LocalValue`. That way the `sink` can bail early when the value no longer exists.

I also have an `identifiedView` which works similarly, but uses the `id` value from underlying `Identified` objects. This can be tricky to manage, and I'm not yet sure that it's necessary in all cases. I need to do some A/B testing to verify my hunch that it would help to improve insertion/removal animations.

**Comment by liscio** [comment]
> The ergonomics of higher-order reducers can get pretty gnarly with free functions, so soon we plan to do an ergonomics episodes upgrading reducers to a proper struct Reducer { reducer: (inout State, Action) -> [Effect<Action>] } type. If you are doing a bunch of higher-order reducer work you may want to try it out

@mbrandonw I have tried this, and it does seem to help.

But I have _another_ suggestion for ergonomics, and it sort of goes in reverse of the above—the aforementioned `Indexed` type should _not_ be a `struct`. Why? _Because **Pattern Matching**_!

Given our hypothetical list of `ProjectCellState` above, pretend it is defined like this:

``` swift
struct ProjectCellState {
  var title: String
  var isSelected: Bool
}

enum ProjectCellAction {
  case toggleSelection
  case setSelected(Bool)
}
```

When someone taps on the project cell, we `self.store.send(.toggleSelection)`. Perhaps the `isSelected` state reveals additional controls.

Now, we might want to allow only a single element to be selected in our `VStack` of `ProjectCellView`s. So how do we achieve this? The `ProjectListState` might look like this:

```swift
struct ProjectListState {
  // Stuff about all the projects, their individual selection states, etc.
}

enum ProjectListAction {
  case cell(Indexed<ProjectCellAction>)
}
```

If we used `struct` for `Indexed`, we'd be stuck with something like this because we can't pattern-match using the embedded struct value:

``` swift
let theReducer = Reducer<ProjectListState, ProjectListAction> { state, action in 
  switch action {
    case let .cell(indexedAction):
      switch indexedAction.value {
        case .toggleSelection:
          let toDeselect = state.projects.indices.filter { $0 != indexedAction.index && state.projectIsSelected(at: $0) }
         return toDeselect.map { otherIndex in Effect { cb in cb(.index(Indexed(index: otherIndex, value: .setSelected(false)))) }
      }
  }
}
```

If we instead use `typealias Indexed<Value> = (index: Int, value: Value)`, we get a much nicer setup:

``` swift
let theReducer = Reducer<ProjectListState, ProjectListAction> { state, action in 
  switch action {
    case let .cell(projectIndex, .toggleSelection):
      let toDeselect = state.projects.indices.filter { $0 != projectIndex && state.projectIsSelected(at: $0) }
      return toDeselect.map { otherIndex in Effect { cb in cb(.index((otherIndex, .setSelected(false))) }
    }
  }
}
```

(Of course, we don't necessarily have to use side-effects to ask children to de-select. I'm just showing an additional benefit of the tuple definition by constructing a child action.)

While this still feels less-than-perfect to "snoop on the child events" to override or counteract their behavior, the problem with the `Indexed` struct gets much, **much worse** as the nesting grows.

Hope this helps!

**Comment by mbrandonw** [comment]
@liscio I haven't had a chance to read everything in this thread yet (lots of great ideas!), but I will say that your destructure idea is good one and we even started doing something similar (but using enums with one case 😂). it's been working out much better since we did that.

**Comment by finestructure** [comment]
This thread has been super helpful trying to sort out the identical problem I ran into. I've got a sample project here: https://github.com/finestructure/PFCompArch which now works – in case someone else wants to have a playground for this.

I've adapted the above mentioned `indexed` reducer to the shape of the `Reducer` encountered later in the series (based on Combine):

```
public func indexed<State, Action, GlobalState, GlobalAction>(
    reducer: @escaping Reducer<State, Action>,
    _ stateKeyPath: WritableKeyPath<GlobalState, [State]>,
    _ actionKeyPath: WritableKeyPath<GlobalAction, Indexed<Action>?>
) -> Reducer<GlobalState, GlobalAction> {
    return { globalValue, globalAction in
        guard let localAction = globalAction[keyPath: actionKeyPath] else { return [] }
        let index = localAction.index
        let localEffects = reducer(&globalValue[keyPath: stateKeyPath][index], localAction.value)
        return localEffects.map { localEffect in
            localEffect.map { localAction in
                var globalAction = globalAction
                globalAction[keyPath: actionKeyPath] = Indexed(index: index, value: localAction)
                return globalAction
            }.eraseToEffect()
        }
    }
}
```

I'll look into adding something like a `detailed` reducer that instead of being based on indexes uses the item directly. I think that should work although when it comes to higher order stuff all my normal intuition of what works and what doesn't seems to break down 😎

On that note, one of the most baffling things about this series is how often the only goal is to make it compile given the inputs. It seems like I don't know 100% what I'm doing but when it compiles I'm done. Back in the day that was when the real problems started 😅 but here often enough it seems to be that there really only is one (sensible) implementation and when it does compile you've kind of stumbled into it.

Is that compile oriented programming? 🤪

**Comment by finestructure** [comment]
I've added a branch `identified` to the [above mentioned repo](https://github.com/finestructure/PFCompArch) which essentially does the same thing as `indexed` but by wrapping an `id` from an `Identifiable` protocol.

The view is then created via

```
            ForEach(store.value.cells) { cell in
                Cell(store: self.store.view(value: { $0.cells.first(where: { $0.id == cell.id })! },
                                            action: { .cell(Identified(id: cell.id, action: $0)) }
                    )
            )}
```

(I suppose for symmetry the `where: ...!` could be cleanup into a subscript `$0.cells[id: cell.id]`)

What I found a little weird about the initial proposal of `indexed` and which this version exposes more prominently is that it's sort of piggy-backing the index - or identifier - into the action, even though it's really tied to the state.

I've tried to make that obvious by defining the "carrying" struct as

```
public struct Identified<Value: Identifiable, Action> {
    public var id: Value.ID
    public var action: Action

    public init(id: Value.ID, action: Action) {
        self.id = id
        self.action = action
    }
}
```

There's nothing tying the second generic to `Action`, this could be named anything. Except `Value`, I guess, because since we use this for `State, Action` pairs we should be honest and reserve `Value` for the `ID` of the first parameter: It's going to be used to identify which `Value` to pick from `[Value]`. (Actually, maybe I should have called the generics `State, Action`.)

`Indexed` glosses over this a bit in the sense that `Int` as an index parameter obscures the fact that the index "belongs" to the array of `Item`s we're looping over. With `Value.ID` it's clearer that we're really attaching the actions to the `Item`'s loop identifiers `index` or `id`.

And that seems a bit like a hack, because in the reducer it's the state projection that uses the `index` or `id` and not (really) the action. Although I guess an argument can be made that it helps identify the `localAction`'s target element.

I hope this makes sense and I would love to hear people's thoughts. Is there a better way to do this? Can `Identified/identified` be improved? Is this even better than just using `Indexed/indexed` at all?

@liscio , does this match your implementation of  `Identified/identifiedView`?

**Comment by finestructure** [comment]
I think I've found a decent workaround for the "delte action from within the list" issue raised above by @liscio .

The observation is a crash when creating the "store view", whether using `Indexed`

```
	self.store.view(value: { $0.cells[idx] },
```

or `Identified`

```
    self.store.view(value: { $0.cells.first(where: { $0.id == item.id })! },
```

The issue is, as mentioned, that the delete action invalidates the list but the view body still loops over the old state.

It doesn't matter, by the way, if the deletion is an action within the detail view (a button in a cell) that propagates up or if it's an `onDelete` modifier

```
    .onDelete { (indexSet) in
        self.store.send(.delete(indexSet))
    }
```

– both cause the same crash.

The simple fix I applied is to essentially to allow the view body to execute successfully no matter what and just inject an `EmptyView` whenever an item is invalid:

```
struct ListView: View {
    @ObservedObject var store: Store<State, Action>

    func cellView(for item: Item) -> AnyView {
        guard let cell = store.value.items.first(where: { $0.id == item.id }) else {
            return AnyView(EmptyView())
        }
        return AnyView(
            ListCell(store: self.store.view(
                value: { _ in .init(item: cell) },
                action: { .cell(Identified(id: cell.id, action: $0)) }))
        )
    }

    var body: some View {
        List {
            ForEach(store.value.items) { cell in
                self.cellView(for: cell)
            }
            .onDelete { (indexSet) in
                self.store.send(.delete(indexSet))
            }
        }
    }
}
```

(full example here: https://github.com/finestructure/PFCompArch/blob/master/PFCompArch/ListTab/ListView.swift)

This seems to work fine and without visual glitches, presumably because the invalid state is not really ever presented before it gets updated again. Or the `EmptyView` just gets ignored.

The `cellView` method could perhaps be turned into some sort of generic `DetailView` and made reusable. Apart from the `Item` and `ListCell` types there's nothing that's specific to my example so my hope is that all this could be tucked away somewhere into a safe, reusable `DetailView<Item, View>`.

**Comment by eivindml** [comment]
Updated `indexed()` to work with CasePaths:

```swift
public func indexed<State, Action, GlobalState, GlobalAction>(
  reducer: @escaping Reducer<State, Action>,
  _ stateKeyPath: WritableKeyPath<GlobalState, [State]>,
  _ actionCasePath: CasePath<GlobalAction, Indexed<Action>>
) -> Reducer<GlobalState, GlobalAction> {
  return { globalValue, globalAction in
    guard let localAction = actionCasePath.extract(from: globalAction) else { return [] }
    let index = localAction.index
    let localEffects = reducer(&globalValue[keyPath: stateKeyPath][index], localAction.value)
    return localEffects.map { localEffect in
      localEffect.map { localAction in
        actionCasePath.embed(Indexed(index: index, value: localAction))
      }.eraseToEffect()
    }
  }
}
```

**Comment by slashmo** [comment]
I created another `view` method on `Store` to simplify viewing into an array of `Identifiable`s:

```swift
public func view<LocalValue: Identifiable, LocalAction>(
    _ array: WritableKeyPath<Value, [LocalValue]>,
    id: LocalValue.ID,
    action: CasePath<Action, Identified<LocalValue, LocalAction>>
) -> Store<LocalValue, LocalAction> {
    view(
        value: { (value: Value) in value[keyPath: array].first(where: { $0.id == id })! },
        action: { action.embed((id, action: $0)) }
    )
}
```

This allows to view into a specific element more easily:

```swift
store.view(\.items, id: 1, action: /Action.item)
```

**Comment by peterkovacs** [comment]
Here is my version of `indexed` which compensates for items being added / removed to the underlying array of state. It requires that the LocalState is Identifiable:

```swift
    public func indexed<LocalState, LocalAction>(
        index: Int,
        value toLocalState: WritableKeyPath<State, [LocalState]>,
        action toGlobalAction: CasePath<Action, (Int, LocalState.ID, LocalAction)>
    ) -> Store<LocalState, LocalAction> where LocalState: Identifiable {
        var index = index
        let localStore =
            Store<LocalState, LocalAction>(
                initialState: self.state[keyPath: toLocalState][index],
                reducer: { (localState, localAction, _) -> [Effect<LocalAction>] in
                    let values = self.state[keyPath: toLocalState]
                    if !values.indices.contains(index) || localState.id != values[index].id {
                        guard let newIndex = values.firstIndex(where: { localState.id == $0.id}) else { return [] }
                        index = newIndex
                    }
                    let globalAction = toGlobalAction.embed((index, localState.id, localAction))

                    self.send(globalAction)
                    localState = self.state[keyPath: toLocalState][index]
                    return []
            },
                environment: self.environment)

        localStore.viewCancellable = self.$state.sink(receiveValue: { [weak localStore] newValue in
            let values = newValue[keyPath: toLocalState]
            if !values.indices.contains(index) || localStore?.state.id != values[index].id {
                guard let newIndex = values.firstIndex(where: { localStore?.state.id == $0.id }) else { return }
                index = newIndex
            }

            localStore?.state = newValue[keyPath: toLocalState][index]
        })

        return localStore
    }
```

**Comment by finestructure** [comment]
@peterkovacs I just wanted to try this out and ended up a little confused 😅

If I understand correctly, this is a `view` similar to what @slashmo has posted above but based on `index` instead of `id`. (We discussed a method `indexed` that's used in reducers upthread, that put me on the wrong track initially.)

However, `id` is also used to find the correct element. That seems redundant - couldn't you make this work with `id` alone? And then it'd be very similar/the same as @slashmo 's version?*) Surely I'm missing something :)

*) What it does in addition though is ensure that values are safely accessed when `sink`ing on the value changes. Seems to me that that could also just do an `id` based filter.
